File: module1/go_impl/airline.go
package main

/*
Extra task
Using trap to get out of the plane
Using terminal to get in the plane
*/
import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type Airplane struct {
	id       int
	capacity int
	rangeVal int
}

func newAirplane(id, capacity, rangeVal int) *Airplane {
	return &Airplane{id, capacity, rangeVal}
}

type Communication struct {
	terminalSemaphore *Semaphore
	trapSemaphore     *Semaphore
}

func newCommunication(terminalCapacity, trapCapacity int) *Communication {
	return &Communication{
		terminalSemaphore: NewSemaphore(terminalCapacity),
		trapSemaphore:     NewSemaphore(trapCapacity),
	}
}

func (c *Communication) communicateTrap() {
	c.communicate(c.trapSemaphore)
}

func (c *Communication) communicateTerminal() {
	c.communicate(c.terminalSemaphore)
}

func (c *Communication) communicate(semaphore *Semaphore) {
	semaphore.acquire()
	time.Sleep(time.Second)
	semaphore.release()
}

type Semaphore struct {
	ch chan int
}

func NewSemaphore(capacity int) *Semaphore {
	return &Semaphore{ch: make(chan int, capacity)}
}

func (s *Semaphore) acquire() {
	s.ch <- 0
}

func (s *Semaphore) release() {
	<-s.ch
}

type Destination struct {
	passengersToTransfer int
	id                   int
	communication        *Communication
	done                 bool
	doneCh               *chan int
	sync                 sync.Mutex
	group                *sync.WaitGroup
}

func NewDestination(id, passengersToTransfer int, communication *Communication, wg *sync.WaitGroup, done *chan int) *Destination {
	return &Destination{
		passengersToTransfer: passengersToTransfer,
		id:                   id,
		communication:        communication,
		group:                wg,
		doneCh:               done,
	}
}

func (d *Destination) serveAirplane(airplane *Airplane) {
	d.communication.communicateTrap()
	if d.done {
		return
	}
	time.Sleep(200 * time.Millisecond)
	d.addPassengers(airplane)
	d.communication.communicateTerminal()
}

func (d *Destination) addPassengers(airplane *Airplane) {
	d.sync.Lock()
	defer d.sync.Unlock()
	toServe := min(d.passengersToTransfer, airplane.capacity)
	d.passengersToTransfer -= toServe
	if d.passengersToTransfer == 0 {
		d.done = true
		d.group.Done()
	}
	fmt.Printf("Plane %d got passengers in destination %d. Passengers left: %d\n", airplane.id, d.id, d.passengersToTransfer)
}

type PlaneRunnable struct {
	airplane        *Airplane
	destinationList []*Destination
	wg              *sync.WaitGroup
}

func newPlaneRunnable(airplane *Airplane, destinationList []*Destination, latch *sync.WaitGroup) *PlaneRunnable {
	return &PlaneRunnable{airplane, destinationList, latch}
}

func (pr *PlaneRunnable) runAirplane() {
	step := pr.airplane.rangeVal
	currentAirline := 0
	for {
		destination := pr.destinationList[currentAirline]
		select {
		case *destination.doneCh <- 0:
			<-*destination.doneCh
		default:
			return
		}
		destination.serveAirplane(pr.airplane)
		currentAirline = (currentAirline + step) % len(pr.destinationList)
	}
}

func main() {

	var destinationList []*Destination
	wg := sync.WaitGroup{}
	wg.Add(7)
	doneChan := make(chan int, 7)
	for i := 0; i < 7; i++ {
		communication := newCommunication(2, 2)
		destinationList = append(destinationList, NewDestination(i, rand.Intn(800-500+1)+500, communication, &wg, &doneChan))
	}

	for i := 0; i < 4; i++ {
		airplane := newAirplane(i, rand.Intn(100)+100, i+1)
		pr := newPlaneRunnable(airplane, destinationList, &wg)
		go pr.runAirplane()
	}

	wg.Wait()
}

File: module1/Module1_Var10/src/main/java/univ/module/Airline.java
package univ.module;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Semaphore;

/**
 * Extra task
 */
public class Airline {
    private static class Airplane {
        private final int id;
        private final int capacity;
        private final int range;

        private Airplane(int id, int capacity, int range) {
            this.id = id;
            this.capacity = capacity;
            this.range = range;
        }

        public int getCapacity() {
            return capacity;
        }

        public int getRange() {
            return range;
        }

        public int getId() {
            return id;
        }
    }

    private static class Communication {
        private final Semaphore terminalSemaphore;
        private final Semaphore trapSemaphore;
        private Communication(int terminalCapacity, int trapCapacity) {
            terminalSemaphore = new Semaphore(terminalCapacity);
            trapSemaphore = new Semaphore(trapCapacity);
        }

        public void communicateTrap() {
            communicate(trapSemaphore);
        }
        public void communicateTerminal() {
            communicate(terminalSemaphore);
        }

        public void communicate(Semaphore semaphore) {
            try {
                semaphore.acquire();
                Thread.sleep(1000);
                semaphore.release();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }
    private static class Destination {
        int passengersToTransfer;
        int id;
        private final Communication communication;

        private boolean done = false;
        private final Object sync = new Object();
        private final CountDownLatch latch;
        public Destination(int id, int passengersToTransfer, Communication communication, CountDownLatch latch) {
            this.passengersToTransfer = passengersToTransfer;
            this.id = id;
            this.communication = communication;
            this.latch = latch;
        }

        private void serveAirplane(Airplane airplane) {
            communication.communicateTrap();
            if (done) {
                return;
            }
            try {
                Thread.sleep(200);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }

            addPassengers(airplane);

            communication.communicateTerminal();
        }

        private void addPassengers(Airplane airplane) {
            synchronized (sync) {
                int toServe = Math.min(passengersToTransfer, airplane.getCapacity());
                passengersToTransfer -= toServe;
                if (passengersToTransfer == 0) {
                    done = true;
                    latch.countDown();
                }
                System.out.println("Plane " + airplane.getId() + " got passengers in destination " + id + ". Passengers left: " + passengersToTransfer);
            }
        }
    }

    private static class PlaneRunnable implements Runnable {

        private final Airplane airplane;
        private final List<Destination> destinationList;
        private final CountDownLatch latch;
        private PlaneRunnable(Airplane airplane, List<Destination> destinationList, CountDownLatch latch) {
            this.airplane = airplane;
            this.destinationList = destinationList;
            this.latch = latch;
        }

        @Override
        public void run() {
            int step = airplane.getRange();
            int currentAirline = 0;
            while (latch.getCount()>0) {
                Destination destination = destinationList.get(currentAirline);
                destination.serveAirplane(airplane);
                currentAirline = (currentAirline + step) % destinationList.size();
            }
        }

    }

    public static void main(String[] args) {
        Random random = new Random();
        List<Destination> destinationList = new ArrayList<>();
        CountDownLatch countDownLatch = new CountDownLatch(7);
        for (int i = 0; i < 7; i++) {
            Communication communication = new Communication(2, 2);
            destinationList.add(new Destination(i, random.nextInt(500, 800), communication, countDownLatch));
        }

        for (int i = 0; i < 4; i++) {
            Airplane airplane = new Airplane(i, random.nextInt(100,200), i+1);
            PlaneRunnable r = new PlaneRunnable(airplane, destinationList, countDownLatch);
            Thread thread1 = new Thread(r);
            thread1.start();
        }

        try {
            countDownLatch.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}
